plutus-core/plutus-core/src/PlutusCore/Default/Builtins.hs

```diff
@@ -1887,89 +1893,90 @@ instance uni ~ DefaultUni => ToBuiltinMeaning uni DefaultFun where
             {-# INLINE xorByteStringDenotation #-}
         in makeBuiltinMeaning
             xorByteStringDenotation
-            (runCostingFunThreeArguments . unimplementedCostingFun)
+            (runCostingFunThreeArguments . paramXorByteString)
 
     toBuiltinMeaning _semvar ComplementByteString =
         let complementByteStringDenotation :: BS.ByteString -> BS.ByteString
             complementByteStringDenotation = Bitwise.complementByteString
             {-# INLINE complementByteStringDenotation #-}
         in makeBuiltinMeaning
             complementByteStringDenotation
-            (runCostingFunOneArgument . unimplementedCostingFun)
+            (runCostingFunOneArgument . paramComplementByteString)
+
+    -- Bitwise operations
 
     toBuiltinMeaning _semvar ReadBit =
         let readBitDenotation :: BS.ByteString -> Int -> BuiltinResult Bool
             readBitDenotation = Bitwise.readBit
             {-# INLINE readBitDenotation #-}
         in makeBuiltinMeaning
             readBitDenotation
-            (runCostingFunTwoArguments . unimplementedCostingFun)
+            (runCostingFunTwoArguments . paramReadBit)
 
     toBuiltinMeaning _semvar WriteBits =
-        let writeBitsDenotation :: BS.ByteString -> [(Integer, Bool)] -> BuiltinResult BS.ByteString
-            writeBitsDenotation = Bitwise.writeBits
+        let writeBitsDenotation :: BS.ByteString -> ListCostedByLength (Integer, Bool) -> BuiltinResult BS.ByteString
+            writeBitsDenotation s (ListCostedByLength updates) = Bitwise.writeBits s updates
             {-# INLINE writeBitsDenotation #-}
         in makeBuiltinMeaning
             writeBitsDenotation
-            (runCostingFunTwoArguments . unimplementedCostingFun)
+            (runCostingFunTwoArguments . paramWriteBits)
 
     toBuiltinMeaning _semvar ReplicateByte =
-        let replicateByteDenotation :: Int -> Word8 -> BuiltinResult BS.ByteString
-            replicateByteDenotation = Bitwise.replicateByte
+        let replicateByteDenotation :: NumBytesCostedAsNumWords -> Word8 -> BuiltinResult BS.ByteString
+            replicateByteDenotation (NumBytesCostedAsNumWords n) w = Bitwise.replicateByte n w
+            -- FIXME: be careful about the coercion in replicateByte
             {-# INLINE replicateByteDenotation #-}
         in makeBuiltinMeaning
             replicateByteDenotation
-            (runCostingFunTwoArguments . unimplementedCostingFun)
-
-    -- Bitwise
+            (runCostingFunTwoArguments . paramReplicateByte)
 
     toBuiltinMeaning _semvar ShiftByteString =
-        let shiftByteStringDenotation :: BS.ByteString -> Int -> BS.ByteString
-            shiftByteStringDenotation = Bitwise.shiftByteString
+        let shiftByteStringDenotation :: BS.ByteString -> IntegerCostedLiterally -> BS.ByteString
+            shiftByteStringDenotation s (IntegerCostedLiterally n) = Bitwise.shiftByteString s (fromIntegral n)
```

Comment:
This is a major bug I believe. `fromIntegral` wraps around:

```
>>> fromIntegral (toInteger (maxBound :: Int) + 1) :: Int
-9223372036854775808
```

and costing won't save you because the builtin is linear in its first argument and the second one is irrelevant costing-wise. Meaning this builtin now accepts large integers, runs and misbehaves without throwing, while previously it'd throw a runtime operational error along the lines of "the given `Integer` isn't in the bounds of `Int`".

And even if this isn't a bug, you can't use `fromIntegral`. Quoting the docs in this module

```
    YOU MUST NEVER USE 'fromIntegral' AND SIMILAR FUNCTIONS THAT CAN SILENTLY UNDER- OR OVERFLOW
    WHEN DEFINING A BUILT-IN FUNCTION
```

It's way too general and can easily result in overflow/underflow if `Int` is changed to `Integer` in the type signature of a builtin like you did with `replicateByte`.

Reply:
Oops! Thanks for pointing that out.  I'll fix it.

Reply:
I think this should be fixed now, but we need to see what happens with #6309 before merging this PR.